src/Core/Services/CipherService.cs

```diff
@@ -571,28 +571,33 @@ public async Task SaveWithServerAsync(Cipher cipher)
             await UpsertAsync(data);
         }
 
-        public async Task ShareWithServerAsync(CipherView cipher, string organizationId, HashSet<string> collectionIds)
+        public async Task ShareWithServerAsync(CipherView cipherView, string organizationId, HashSet<string> collectionIds)
         {
             var attachmentTasks = new List<Task>();
-            if (cipher.Attachments != null)
+            Cipher cipher = null;
+            //If the cipher doesn't have a key, we update it 
+            if(cipherView.Key == null)
             {
-                foreach (var attachment in cipher.Attachments)
+                cipher = await EncryptAsync(cipherView);
+                await UpdateAndUpsertAsync(async () => await _apiService.PutCipherAsync(cipherView.Id, new CipherRequest(cipher)));
+                cipher = await GetAsync(cipherView.Id);
+                cipherView = await cipher.DecryptAsync();
+            }
+            if (cipherView.Attachments != null)
+            {
+                foreach (var attachment in cipherView.Attachments)
                 {
                     if (attachment.Key == null)
                     {
-                        attachmentTasks.Add(ShareAttachmentWithServerAsync(attachment, cipher.Id, organizationId));
+                        attachmentTasks.Add(ShareAttachmentWithServerAsync(attachment, cipherView.Id, organizationId));
                     }
                 }
             }
             await Task.WhenAll(attachmentTasks);
-            cipher.OrganizationId = organizationId;
-            cipher.CollectionIds = collectionIds;
-            var encCipher = await EncryptAsync(cipher);
-            var request = new CipherShareRequest(encCipher);
-            var response = await _apiService.PutShareCipherAsync(cipher.Id, request);
-            var userId = await _stateService.GetActiveUserIdAsync();
-            var data = new CipherData(response, userId, collectionIds);
-            await UpsertAsync(data);
+            cipherView.OrganizationId = organizationId;
+            cipherView.CollectionIds = collectionIds;
+            cipher = await EncryptAsync(cipherView);
```

Comment:
I didn't do that because I need the cipher to create two different requests, CipherRequest and CipherShareRequest. I could add in the api call but I think there's already too much happening in one line of code.
`await UpdateAndUpsertAsync(async () => await _apiService.PutShareCipherAsync(cipherView.Id, new CipherShareRequest(await EncryptAsync(cipherView))), collectionIds);`

Reply:
Oh I see, I think it's fine to pass the `CipherView` as a parameter to the func to construct such requests.